### PR TITLE
Fix a deadlock in ShowWindow

### LINKED ISSUE #13301: Iconify window op causes conhost to hang
### Windows Terminal version

Commit 9eb191d5453fed890bad63952e425e1240aa8df5

### Windows build number

10.0.19044.1706

### Other Software

_No response_

### Steps to reproduce

1. Build a recent version of OpenConsole.
2. Open a bash shell.
3. Execute `printf "\e[2t"`

### Expected Behavior

The console window should be minimized.

### Actual Behavior

The window is briefly minimized and then restored again, but at that point it's no longer responding. Can't move it. Can't close it. Have to go to taskman to kill it.

When run from within the debugger, it appears to be blocking inside the [`ShowWindow`](https://github.com/microsoft/terminal/blob/799b5d4add799ac2fbf5c855b774cd7dcb957697/src/host/outputStream.cpp#L289) call, somewhere in "External Code".

### PR DESCRIPTION
**Pull request by zadjii-msft**
When we send this ShowWindow message, if we send it to it's
going to need to get processed by the window message thread before
returning. We're handling this message under lock. However, the first
thing the conhost message thread does is lock the console. That'll
deadlock us. So unlock here, first, to let the message thread deal with
this message, then re-lock so later on this thread can unlock again
safely.

* [x] Closes #13301 
* [x] Tested conhost
* [x] Tested terminal

### CODE REVIEW

**Comment by zadjii-msft** [comment]
notes: also look into `PostMessage`

**Comment by zadjii-msft** [comment]
More astutely: `::PostMessage(hwnd, WM_SYSCOMMAND, showOrHide ? SC_RESTORE : SC_MINIMIZE, 0);` does NOT work, because it's got the activate-the-conpty-window on restore issue

File: src/host/outputStream.cpp [hunk 1]
```diff
@@ -286,7 +286,17 @@ void ConhostInternalGetSet::ShowWindow(bool showOrHide)
 {
     const auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
     const auto hwnd = gci.IsInVtIoMode() ? ServiceLocator::LocatePseudoWindow() : ServiceLocator::LocateConsoleWindow()->GetWindowHandle();
+
+    // GH#13301 - When we send this ShowWindow message, if we send it to it's
+    // going to need to get processed by the window message thread before
```
Comment (carlos-zamora) [inline src/host/outputStream.cpp:291]:
```
    // GH#13301 - When we send this ShowWindow message, if we send it to it's
    // going to need to get processed by the window message thread before
```

send it to what? I can't parse this comment haha.

Comment (zadjii-msft) [reply src/host/outputStream.cpp:291]: guh I don't even know. Lemme rewrite that

**Review by carlos-zamora — APPROVED** [review]
Approving but I'm guessing @DHowett may want to take a look? 🤷‍♂️

File: src/host/outputStream.cpp [hunk 2]
```diff
@@ -286,7 +286,17 @@ void ConhostInternalGetSet::ShowWindow(bool showOrHide)
...
+    // GH#13301 - When we send this ShowWindow message, if we send it to the
+    // conhost HWND, it's going to need to get processed by the window message
+    // thread before returning. We (the VT adapter) are handling this message
+    // under lock. However, the first thing the conhost message thread does is
+    // lock the console. That'll deadlock us. So unlock here, first, to let the
+    // message thread deal with this message, then re-lock so later on this
+    // thread can unlock again safely.
+    gci.UnlockConsole();
     ::ShowWindow(hwnd, showOrHide ? SW_SHOWNOACTIVATE : SW_MINIMIZE);
+    gci.LockConsole();
```
Comment (lhecker) [inline src/host/outputStream.cpp:299]: We did something similar for MIDI recently. So I think this is fine.
Do we need to wait for `ShowWindow` to complete? If not `ShowWindowAsync` can be used instead.
BTW I'm surprised this doesn't use `SW_RESTORE`.

Comment (zadjii-msft) [reply src/host/outputStream.cpp:299]: what-000

Comment (zadjii-msft) [reply src/host/outputStream.cpp:299]: I literally till this moment did not know there was a `ShowWindowAsync`. That needs to be SO MUCH HIGHER in that doc.

**Review by lhecker — APPROVED** [review]

**Comment by DHowett** [comment]
> BTW I'm surprised this doesn't use SW_RESTORE.

I am too, but I also bet there's a big and important reason.

**Pull request merged**